### PR TITLE
Handle attributes disabled for variations usage

### LINKED ISSUE #3769: [Variations] Handle Variations with no Attribute set available
If a Variable Product gets all available attributes removed and leave every Variation under it with no Attribute set, the Variation List ends showing a list of Variations with no title and the Attributes editing becomes completely inoperative, we should filter Variations with no Attribute set as invalid ones

image

### LINKED ISSUE #3770: [Variations] Variation Attribute PUT request returns HTTP 200 but no change is done under the app
### ⚠️ Warning: There's a chance this is an issue with Woo Core, but further investigation is required to understand what's happening

Opening a variation that has "Color: Any" then changed the color to "Blue" and tapped the Update button. The request succeeds but the color changes back to "Any."

Looking at the network traffic, the request body seems fine:

```
body: "{"image":{"id":3365,"date_created":"","src":"https://i1.wp.com/dextercomicstrip.com/wp-content/uploads/2021/03/PXL_20210316_151559874.jpg?fit\u003d4032%2C3024\u0026ssl\u003d1","alt":"","name":"PXL_20210316_151559874.jpg"},
"attributes":[{"id":5,"name":"Color","option":"Blue"}]}"
 json: "true"
 path: "/wc/v3/products/2926/variations/3357/&_method=put"
```

But even though the request gets an HTTP 200 response, the response body shows `attributes: []`.

ezgif-2-1e2c5eae47f2

### LINKED ISSUE #3771: [Variations] Attributes with the `use for variation` flag set as false should be ignored as selectable
_Pending description_

### LINKED ISSUE #3913: [Variations] Product Attributes disabled for variations should be re-enabled from the `Edit Attributes` flow
When adding attributes to a Variable Product, we can choose from a list of Global Attributes with no issues, but we can also list Attributes from that Product who aren't enabled for variations usage through the `use for variations` flag, to avoid making the merchant going to the store to re-enable the Product Attribute, we could improve the `AddAttributeFragment` to support the Product Attribute reconfiguration.

### PR DESCRIPTION
**Pull request by ThomazFB**
Summary
==========
Fixes issues #3771, #3770, #3913 and #3769 by creating filters to only enable for Variations attributes set as true for the `use for variation` flag. Since issue #3770 is not a bug from the API, but a consequence of trying to update inside a Variation attributes who aren't enabled for variation usage, this PR won't solve the issue itself but block the described scenario from happening.

How to test
==========
Before testing these changes, make sure to set up a Variable Product with Attributes flagged as both `true` and `false` for `use for variations`, so we can properly reproduce a scenario with Attributes to be filtered out.

Make sure that for every Variation view, only the enabled for variation usage are displayed (except the Add Attributes view, since that would be view where we want to add Attributes who are disabled for variations)

image

# Before
| Product Detail | Variation List | Variation Detail | Edit Variation Attributes | Attributes List |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Screenshot_20210426_220719 | Screenshot_20210426_220723 | Screenshot_20210426_220727 | Screenshot_20210426_220738 | Screenshot_20210426_220749 |

# After
| Product Detail | Variation List | Variation Detail | Edit Variation Attributes | Attributes List |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Screenshot_20210426_215900 | Screenshot_20210426_215845 | Screenshot_20210426_215851 | Screenshot_20210426_215854 | Screenshot_20210426_221049 |

Update release notes:

- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by nbradbury** [comment]
I'm a bit confused by this. I have a variable product with two global attributes, Color and Shape, and local attribute Size. "Use for variations" is enabled for Color and Size, but not for Shape.

Screen Shot 2021-04-27 at 7 14 10 AM

When I select "Edit attributes" from the variation list I see only Color and Size as attributes (Shape isn't available). But when I click "Add attribute" I still see Shape as an available attribute.



Is this how it should work?

Also, and perhaps this is a separate issue, when I tap "Add variation," the request body shows that Shape is still being passed, even though "Use for variations" isn't enabled for it.

```
body: "{"attributes":[{"id":5,"name":"Color","option":""},
{"id":0,"name":"Shape","option":""},
{"id":0,"name":"size","option":""}]}"
 json: "true"
 path: "/wc/v3/products/2926/variations/&_method=post"
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt [hunk 1]
```diff
@@ -259,6 +261,12 @@ data class Product(
         return true
     }
 
+    infix fun containsValidAttributesFor(variation: ProductVariation) =
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/Product.kt]: **TIL**: "infix"

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/ProductVariation.kt [hunk 2]
```diff
@@ -147,12 +147,23 @@ data class ProductVariation(
...
 
     fun getName(parentProduct: Product? = null): String {
-        return parentProduct?.attributes?.joinToString(" - ") { attribute ->
+        return parentProduct?.variationEnabledAttributes?.joinToString(" - ") { attribute ->
             val option = options.firstOrNull { it.attributeName == attribute.name }
             option?.optionChoice ?: "Any ${attribute.name}"
         } ?: options.joinToString(" - ") { o -> o.optionChoice }
     }
 
+    fun containsValidAttributesFrom(product: Product) =
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/model/ProductVariation.kt]: This and `containsValidAttributesFor` might benefit from a brief comment explaining their purpose.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 3]
```diff
@@ -194,6 +194,18 @@ class ProductDetailViewModel @AssistedInject constructor(
...
     }
 
+    /**
+     * Returns the list of attributes assigned to the product who are enabled for Variations
+     */
+    val productDraftVariationAttributes
+        get() = viewState.productDraft?.variationEnabledAttributes ?: emptyList()
+
+    /**
+     * Returns the list of attributes assigned to the product who are disabled for Variations
```
Comment (nbradbury) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Is this comment correct? It looks like this returns the entire list of attributes, not just ones disabled for variations.

Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: Nice catch, at first this property was just for the disabled attributes, but I changed my mind later and forgot to update the comment. Here's the change: https://github.com/woocommerce/woocommerce-android/pull/3912/commits/499255f0b144214562f9797fc8fe8388ea6fdc64

File: WooCommerce/src/main/kotlin/com/woocommerce/android/model/ProductVariation.kt (hunk 2 above)
Comment (ThomazFB) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/model/ProductVariation.kt]: I agree! But after some thinking around these two methods, I noticed there's a better way to filter variations without the need of creating them, so they're now deleted. Here's the change: https://github.com/woocommerce/woocommerce-android/pull/3912/commits/9e12cbbb6a1553656391e7da68f85ea0f961afba

**Comment by ThomazFB** [comment]
>When I select "Edit attributes" from the variation list I see only Color and Size as attributes (Shape isn't available). But when I click "Add attribute" I still see Shape as an available attribute.

This was intended on my part, and it's related to issue #3913 (one of those solved by this PR), the `tap to select an existing attribute` section at the Add Attribute View is showing both Global Attributes and Local Attributes with the `use for variations` disabled, this way we can also allow the merchant to re-enable the Local Product Attributes for variations usage since this flow is inside the Product Variation section. Without this, we would be requiring the merchant to go to the site in order to do so, and even lock the entire Variation section from the app in scenarios where all Local Attributes got deselected as `use for variations`. WDYT about that? Does it make sense?

>Also, and perhaps this is a separate issue, when I tap "Add variation," the request body shows that Shape is still being passed, even though "Use for variations" isn't enabled for it.

Oops, really good catch on this one! In order to fix this unnecessary data inside request, it will require a FluxC fix at the `WCProductStore.generateEmptyVariation` method, the change is available here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1974, and you can test the change directly from this PR since I've updated the code with the FluxC change right here: 4641f578567a255dedb085df71d3267f44b3f89a.

**Comment by ThomazFB** [comment]
Hey @nbradbury, thanks for your great review! The PR is ready for a second round 😄

**Comment by nbradbury** [comment]
> Without this, we would be requiring the merchant to go to the site in order to do so, and even lock the entire Variation section from the app in scenarios where all Local Attributes got deselected as use for variations. WDYT about that? Does it make sense?

Yep, that makes sense, and it turns out iOS is doing the same thing.

**Review by nbradbury — APPROVED** [review]
:shipit:

**Pull request merged**